src/order/complete_partial_order.lean

```diff
@@ -0,0 +1,391 @@
+/- complete partial orders
+Helpful for a small-scale domain theory, see
+  http://home.in.tum.de/~hoelzl/documents/lochbihler2014recursive.pdf
+-/
+
+import data.pfun
+import data.stream.basic
+import tactic.wlog
+import tactic.find_unused
+
+universes u v
+
+open_locale classical
+
+section mono
+variables {α : Type*} {β : Type*}
+variables  [preorder α]
+
+lemma const_mono [preorder β] (f : β) : monotone (λ _ : α, f) :=
+assume x y h, le_refl _
+
+lemma ite_mono [preorder β] {f g : α → β}
+  {p : Prop} {h : decidable p}
+  (hf : monotone f) (hg : monotone g) :
+  monotone (λ x, @ite _ h _ (f x) (g x)) :=
+by intros x y h; dsimp; split_ifs; [apply hf h, apply hg h]
+
+lemma bind_mono {β γ} (f : α → roption β)
+                (g : α → β → roption γ)
+  (hf : monotone f) (hg : monotone g) :
+  monotone (λ x, f x >>= g x)  :=
+begin
+  intros x y h a, simp, intros b hb ha,
+  refine ⟨b,hf h _ hb,hg h _ _ ha⟩,
+end
+
+end mono
+
+
+/-- Chains are monotonically increasing sequences -/
+structure chain (α : Type u) [preorder α] :=
```

Comment:
This terminology clashes with the naming in `zorn`, and `Sup` clashes with the `Sup` for complete lattices. Is there a way to unify these, or is computability of the functions relevant here?

Reply:
It's important to take them as a series because otherwise `roption` cannot have `Sup` function. As for name clashes, I think we need to use namespaces.

Reply:
This is also only countably complete, since it uses sequences.

Reply:
I can add `omega` to the name.